src/recon_buildblock/PLSPrior.cxx

```diff
@@ -0,0 +1,874 @@
+//
+//
+/*
+    Copyright (C) 2000- 2011, Hammersmith Imanet Ltd
+    This file is part of STIR.
+
+    This file is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.
+
+    This file is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+
+    See STIR/LICENSE.txt for details
+*/
+/*!
+  \file
+  \ingroup priors
+  \brief  implementation of the stir::PLSPrior class
+
+  \author Daniel Deidda
+  \author Tsai Yu Jung
+
+*/
+
+#include "stir/recon_buildblock/PLSPrior.h"
+#include "stir/Succeeded.h"
+#include "stir/DiscretisedDensityOnCartesianGrid.h"
+#include "stir/IndexRange3D.h"
+#include "stir/IO///write_to_file.h"
+#include "stir/IO/read_from_file.h"
+#include "stir/is_null_ptr.h"
+#include "stir/info.h"
+#include <algorithm>
+using std::min;
+using std::max;
+
+/* Pretty horrible code because we don't have an iterator of neigbhourhoods yet
+ */
+
+START_NAMESPACE_STIR
+
+template <typename elemT>
+void
+PLSPrior<elemT>::initialise_keymap()
+{
+  base_type::initialise_keymap();
+  this->parser.add_start_key("PLS Prior Parameters");
+  this->parser.add_key("only 2D", &only_2D);
+  this->parser.add_key("scale_par", &scale_par);
+  this->parser.add_key("smooth_par", &smooth_par);
+  this->parser.add_key("kappa filename", &kappa_filename);
+  this->parser.add_key("anatomical_filename", &anatomical_filename);
+  this->parser.add_key("weights", &weights);
+  this->parser.add_key("gradient filename prefix", &gradient_filename_prefix);
+  this->parser.add_stop_key("END PLS Prior Parameters");
+}
+
+
+template <typename elemT>
+bool
+PLSPrior<elemT>::post_processing()
+{
+  if (base_type::post_processing()==true)
+    return true;
+  if (kappa_filename.size() != 0)
+    this->kappa_ptr = read_from_file<DiscretisedDensity<3,elemT> >(kappa_filename);
+
+  if (anatomical_filename.size() != 0){
+    this->anatomical_sptr = read_from_file<DiscretisedDensity<3,elemT> >(anatomical_filename);}
+  else{
+      this->anatomical_sptr->fill (0);
+  }
+
+
+  if (this->anatomical_filename != "0"){
+      info(boost::format("Reading MR data '%1%'")
+           % anatomical_filename  );}
+
+  bool warn_about_even_size = false;
+
+  DiscretisedDensity<3,elemT> &anat_im_grad_z=*this->anatomical_sptr.get ()->get_empty_copy ();
```

Comment:
this is very dangerous for memory leaks or worse. Use something like this
```c++
shared_ptr<DiscretisedDensity<3,elemT> > px(this->anatomical_sptr->clone());
```
Of course, you should be able to immediatelly assign it as well
```c++
this->anat_grad_x_sptr.reset(this->anatomical_sptr->clone());
```
and just use `*this->anat_grad_x_sptr` below.

Reply:
what if I want to define an image object with all voxels equal to zero? if I use clone() it will start from the actual value pointed by anatomical_sptr. Right?

Reply:
yes true. can use `get_empty_copy()` as well in the same way as above. both return a pointer to a newly allocated object.

Reply:
Ok, so if I correctly understand I should not define the image, im, directly but through pointers, im_sptr, and when I need to call the value  of im[z][y][x] can I do the following?
*im_sptr[z][y][x]?

Reply:
yes, although you might have to write `(*im_sptr)[z][y][x]`. If you want, you can make the notation a bit lighter but create a reference
```c++
DiscretisedDensity<3,elemT> &anat_im_grad_z = *anat_im_grad_z_sptr;
```
This is safe, as it doesn't create a new image. `anat_im_grad_z` is just a reference to an existing image, no allocations, copying etc